File: app/services/exportacao_csv_service.py
"""Exportação CSV do Dashboard de Plantio (feature 015).

`data-model.md` desta feature lista "umidade 0-7cm (%)" como coluna — esse
valor não é uma medição real persistida por talhão (não existe endpoint nem
tabela para ele; é usado apenas como entrada transiente do Balanço Hídrico,
feature 010, vindo da previsão do Open-Meteo para o instante do cálculo).
Substituído aqui por `armazenamento_mm`/`percentual_cad`, que são os valores
reais e persistidos do Balanço Hídrico Diário (mesma fonte do Dashboard,
feature 011) — nunca exportar um número que pareça medição real sem ser.
"""

import csv
import io

CABECALHOS = ["Propriedade", "Talhao", "Area (ha)", "Solo", "Status", "Armazenamento (mm)", "% da CAD"]


def gerar_csv_talhoes(itens: list[dict]) -> str:
    buffer = io.StringIO()
    escritor = csv.writer(buffer, delimiter=";")
    escritor.writerow(CABECALHOS)
    for item in itens:
        escritor.writerow(
            [
                item["propriedade"],
                item["nome"],
                f"{item['area_ha']:.1f}",
                item["tipo_solo"] or "",
                item["status_plantio"] or "SEM_CALCULO",
                f"{item['armazenamento_mm']:.1f}" if item["armazenamento_mm"] is not None else "",
                f"{item['percentual_cad']:.1f}" if item["percentual_cad"] is not None else "",
            ]
        )
    # BOM UTF-8 (FR-003) — sem ele, o Excel em português interpreta o arquivo
    # como Latin-1 e corrompe a acentuação.
    return "﻿" + buffer.getvalue()

File: src/main/java/com/agroclima/api/business/dashboard/ExportacaoCsvService.java
package com.agroclima.api.business.dashboard;

import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Locale;

/** Espelha app/services/exportacao_csv_service.py (feature 015). Delimitador ";" e BOM UTF-8 (convencao Excel pt-BR). */
@Service
public class ExportacaoCsvService {

    private static final String BOM = "﻿";
    private static final String[] CABECALHO = {
        "Propriedade", "Talhao", "Area (ha)", "Solo", "Status", "Armazenamento (mm)", "% da CAD"
    };

    public String gerarCsvTalhoes(List<DashboardItemProjecao> itens) {
        StringBuilder csv = new StringBuilder(BOM);
        csv.append(String.join(";", CABECALHO)).append("\n");
        for (DashboardItemProjecao item : itens) {
            double areaHa = item.getAreaHa() != null ? item.getAreaHa() : 0.0;
            String solo = item.getTipoSolo() != null ? item.getTipoSolo() : "";
            String status = item.getStatusPlantio() != null ? item.getStatusPlantio() : "SEM_CALCULO";
            String armazenamento = item.getArmazenamentoMm() != null ? formatar1Casa(item.getArmazenamentoMm()) : "";
            String percentualCad = (item.getArmazenamentoMm() != null && item.getCadMm() != null && item.getCadMm() > 0)
                    ? formatar1Casa(item.getArmazenamentoMm() / item.getCadMm() * 100.0)
                    : "";
            csv.append(String.join(";",
                            item.getPropriedadeNome(),
                            item.getNome(),
                            formatar1Casa(areaHa),
                            solo,
                            status,
                            armazenamento,
                            percentualCad))
                    .append("\n");
        }
        return csv.toString();
    }

    private String formatar1Casa(double valor) {
        return String.format(Locale.US, "%.1f", valor);
    }
}
